src/backtrace.rs

```diff
@@ -0,0 +1,77 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use core::arch::asm;
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturn {
+    Some(usize),
+    None,
+    BadFp(usize),
+    BadAddress(usize),
+}
+
+pub(crate) fn backtrace() -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    unsafe { backtrace_internal(fp) }
+}
+
+unsafe fn backtrace_internal(fp: u64) -> [BTReturn; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [BTReturn::None; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    loop {
+        // Safe because we are only reading in the valid stack trace area
```

Comment:
Nit: Comment should be moved a few lines below.

Reply:
fixed